### PR TITLE
improve batching behaviour for DynoDM

### PR DESCRIPTION
**Pull request by lueue**

### CODE REVIEW

File: services/aws/dynamodb/DynamoDataStorage.ts [hunk 1]
```diff
@@ -78,9 +105,26 @@ abstract class DynamoDataStorage<TEntity>
     const dynamoQuery: DynamoDB.DocumentClient.QueryInput =
       QueryBuilder.buildQuery(query, this.tableName, this.keys)
 
-    const results = await this.documentClient.query(dynamoQuery).promise()
+    const results = []
+    let lastEvaluatedKey
+    do {
+      const page: DynamoDB.DocumentClient.QueryOutput =
+        // eslint-disable-next-line no-await-in-loop
+        await this.documentClient
```
Comment (fean) [inline services/aws/dynamodb/DynamoDataStorage.ts:113]:
```suggestion
      if (page.Items?.length) {
```

File: services/aws/dynamodb/DynamoDataStorage.ts [hunk 2]
```diff
@@ -12,6 +13,32 @@ abstract class DynamoDataStorage<TEntity>
...
+    }
+
+    const batches = []
+    while (writeRequests.length) {
+      batches.push(writeRequests.splice(0, batchSize))
+    }
+
+    await Promise.all(
+      batches.map((batch) =>
+        this.documentClient
```
Comment (fean) [inline services/aws/dynamodb/DynamoDataStorage.ts:31]: Now what happens if there is a duplicate key inside one of the batches? 
Think of the following scenario: Three batches were written one batch failed due to a key constraint. 

How do you enforce that the data source is going to be consistent with the application state? 🙂

**Review by fean — CHANGES_REQUESTED** [review]

File: services/aws/dynamodb/DynamoDataStorage.ts (hunk 2 above)
Comment (lueue) [reply services/aws/dynamodb/DynamoDataStorage.ts]: I added success and failed states for the items to keep track of what is going on or which items failed. Then the implementor can decide what needs to happen

File: services/aws/dynamodb/DynamoDataStorage.ts [hunk 3]
```diff
@@ -12,6 +15,65 @@ abstract class DynamoDataStorage<TEntity>
...
+          .batchWrite({
+            RequestItems: {
+              [this.tableName]: keyPairs,
+            },
+          })
+          .promise()
+
+        success.push(...batch)
+      } catch (error) {
+        failed.push(...batch)
```
Comment (fean) [inline services/aws/dynamodb/DynamoDataStorage.ts:65]: Are you sure this is the right way to process those items?
 [Check out the documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html#batchWrite-property) for `batchWrite`. In the returning object there is a property `UnprocessedItems`, now I don't know in which cases that property is used. So maybe you could look into that as well? 🙂

**Review by fean — CHANGES_REQUESTED** [review]

**Comment by fean** [comment]
Awesome addition man! Looks good! 💪

**Review by fean — APPROVED** [review]

**Pull request merged**